pennylane/optimize/rotosolve.py

```diff
@@ -132,340 +212,410 @@ class RotosolveOptimizer:
+    Cost before step: -1.0
+    Minimization substeps: [-1. -1. -1. -1. -1. -1. -1.]
+
+    As we can see, while the optimization got a bit harder and the optimizer takes a bit longer
+    to converge than previously, Rotosolve was able to adapt to the more complicated
+    dependence on the input arguments and still found the global minimum successfully.
     """
     # pylint: disable=too-few-public-methods
 
+    def __init__(self, optimizer=None, optimizer_kwargs=None):
+        self.optimizer, self.optimizer_kwargs = _prepare_optimizer(optimizer, optimizer_kwargs)
+
+    def _validate_inputs(self, requires_grad, args, nums_frequency, spectra):
+        """Checks that for each trainable argument either the number of
+        frequencies or the frequency spectrum is given."""
+
+        for arg, (arg_name, _requires_grad) in zip(args, requires_grad.items()):
+            if _requires_grad:
+                _nums_frequency = nums_frequency.get(arg_name, {})
+                _spectra = spectra.get(arg_name, {})
+                all_keys = set(_nums_frequency) | set(_spectra)
+
+                shape = qml.math.shape(arg)
+                indices = np.ndindex(shape) if len(shape) > 0 else [()]
+                for par_idx in indices:
+                    if par_idx not in all_keys:
+                        raise ValueError(
+                            "Neither the number of frequencies nor the frequency spectrum "
+                            f"was provided for the entry {par_idx} of argument {arg_name}."
+                        )
+
     def step_and_cost(
         self,
         objective_fn,
         *args,
-        num_freqs=None,
-        optimizer=None,
-        optimizer_kwargs=None,
+        nums_frequency=None,
+        spectra=None,
+        shifts=None,
         full_output=False,
         **kwargs,
     ):
         r"""Update args with one step of the optimizer and return the corresponding objective
-        function value prior to the step.
+        function value prior to the step. Each step includes multiple substeps, one per
+        parameter.
 
         Args:
             objective_fn (function): the objective function for optimization. It should take a
                 sequence of the values ``*args`` and a list of the gates ``generators`` as inputs,
                 and return a single value.
-            *args : variable length sequence containing the initial
-                values of the variables to be optimized over or a single float with the initial
-                value.
-            num_freqs (int or array[int]): The number of frequencies in the ``objective_fn`` per
-                parameter. If an ``int``, the same number is used for all parameters; if
-                ``array[int]``, the shape of ``args`` and ``num_freqs`` has to coincide.
-                Defaults to ``num_freqs=1``, corresponding to Pauli rotation gates.
-            optimizer (callable or str): the optimization method used for the univariate
```

Comment:
moving this to `init` is also a breaking change

Reply:
Yes! :grimacing: